### ISSUE TITLE
Enhancement: Global transaction log

### ISSUE THREAD

**Issue by FlorianMickler** [syncthing/syncthing#2860]
What I would love is a global transaction log which is synced between all nodes and which contains the global view of a Share (transaction wise). 
As soon as a node goes online and finds a node with that Share, it syncs it's view of the global state with that node. 
I guess this is already happening implicitly in the protocol, but it would be nice to know which node introduced what file and deleted that file and how that file got these modifications (-> through another node which was only online for a short time). 

I hope this wasn't already discussed or proposed in another issue but I didn't find anything resembling this.

**Comment by AudriusButkevicius** [comment]
We only have the last state, ala everyone's view of the world, and we have no clue who made the last modification.

Yet I understand why it would be nice to have this.

**Comment by nrm21** [comment]
I did try to do something like this myself, only to abandon my 2 weekends of coding after realizing it was WAY harder than I originally thought it would be.

Theoretically it should be possible for some node to realize that it has a higher version of a file than the globally known one and that it seems to have the only copy of it.  Then it could attach it's device name to some "global" log saying it created/modified a file at such and such time... then send the file out to others.

**Comment by nrm21** [comment]
Ok what about this:

-----
{"id":85653,"time":"2016-04-09T23:10:39.9833416-04:00","type":"**LocalIndexUpdated**","data":{"filenames":["Test.txt"],"folder":"Sync","items":1,"version":11808}}
{"id":85654,"time":"2016-04-09T23:10:39.9833416-04:00","type":"StateChanged","data":{"duration":0.0156247,"folder":"Sync","from":"scanning","to":"idle"}}
{"id":85655,"time":"2016-04-09T23:10:40.1708865-04:00","type":"**RemoteIndexUpdated**","data":{"device":"TW54YE7-QW5BOZD-JXYUESG-6W5GR6D-ZQH7MRT-XEH55MZ-PFKSDTX-XXXXXXX","folder":"Sync","items":1,"version":9407}}
{"id":85656,"time":"2016-04-09T23:10:40.1708865-04:00","type":"**StateChanged**","data":{"duration":0.1875449,"folder":"Sync","from":"idle","to":"syncing"}}
{"id":85657,"time":"2016-04-09T23:10:40.1708865-04:00","type":"StateChanged","data":{"duration":0,"folder":"Sync","from":"syncing","to":"idle"}}
{"id":85658,"time":"2016-04-09T23:10:40.1708865-04:00","type":"FolderSummary","data":{"folder":"Sync","summary":{"globalBytes":43335289,"globalDeleted":3,"globalFiles":60,"ignorePatterns":false,"inSyncBytes":43335289,"inSyncFiles":60,"invalid":"","localBytes":43335289,"localDeleted":3,"localFiles":60,"needBytes":0,"needFiles":0,"state":"idle","stateChanged":"2016-04-09T23:10:40.1708865-04:00","version":21215}}}
{"id":85659,"time":"2016-04-09T23:10:40.1708865-04:00","type":"FolderCompletion","data":{"completion":100,"device":"TW54YE7-QW5BOZD-JXYUESG-6W5GR6D-ZQH7MRT-XEH55MZ-PFKSDTX-XXXXXXX","folder":"Sync"}}

-----
It looks like whenever a client creates or updates a file it seems to always call **LocalIndexUpdated** then **RemoteIndexUpdated** then **StateChanged ("from":"idle","to":"syncing")** in that order (and in relatively short time from each other).

Wouldn't it seem logical that if one were to create some code that checked for those event's in that order then it could be understood that the local client was the one to first write the file?  And if the order were broken it would reset to the beginning (a simple state machine if you will).

I've already figured out that if a machine is sent a delete/update to a file folder, then the log shows **RemoteIndexUpdated** then **LocalIndexUpdated** but never the other way around.

I would like to run this by you guys before starting on coding something like this.

**Comment by AudriusButkevicius** [comment]
All of this you are talking about needs you to actually be online.

**Comment by nrm21** [comment]
Well I'm online right now.   (lol j/k I know what you meant)

Ok so you're saying that if a ST client is offline makes a file change, and then comes online, we cannot guarantee the order will be the same?

**Comment by nrm21** [comment]
> All of this you are talking about needs you to actually be online.

Ok then maybe a compromise... the log would only include files where the changes were made by online clients.

Not perfect but as long as users are made aware of the limitation.  And besides I'd think it could be a good starting point where we could go back later and figure out how to log offline file mods too after getting working code.

**Comment by calmh** [comment]
Maybe try something simpler, like hooking into the scanner (that detects all local changes), and making it log what you're interested in. It could log that to a local file per device. Sync these files if you like (although you'll have to exclude changes to the log from the log or it'll become circular). Then you can merge these files for a global view. Or you could pull the events from the REST API to a central place.

**Comment by nrm21** [comment]
Yeah I had thought of this first but actually though it might be more complex not simpler.  My thoughts were that having to rectify who wrote what when merging the logs would be more complex code (that could potentially break more things).

And of course using the program to sync a file among the devices is what I planned instead of reinventing the wheel  And yes changes to log triggering more sync, etc, in an endless loop is an issue that would need to be avoided no matter what way it were coded.

**Comment by nrm21** [comment]
Maybe it's possible to set a particular file to sync outside of the configured scan directories.  Like in %APPDATA%/Local/Syncthing/Global.log (that path on Windows anyway) where the audit trail is kept.  Since it's outside of the "scan zone".  Changes could be made to it at some interval and not worry about the circular issue.

**Comment by calmh** [comment]
Did you look at the audit log (running with -audit)? It could be machine merged.

**Comment by nrm21** [comment]
I believe I did.  Wasn't aware it could be merged though.  Can you elaborate?

**Comment by calmh** [comment]
Not really? I mean it's just a file, with one event per line, ordered chronologically. If any of the existing event types match what you need, filter for that and combine with the files from other devices. Or we can add an event type, I'd probably rather do that than the special magic suggested up thread. :)

**Comment by FlorianMickler** [comment]
What about just using a git Repo for that global state? Maybe it could be integrated a little bit into syncthing,  but the package-database seems to be ideal for syncthing and every Scan that detected local changes would just commit the changes and then send everything out to the other Nodes. 
Incoming Transfers would be merged and pushed out again...  That way eventually all would end up with a complete global Log alongside their data. 
Maybe one could turn it on or off depending on what would be synced. 



Am 11. April 2016 21:33:36 MESZ, schrieb Jakob Borg <notifications@github.com>:
>Not really? I mean it's just a file, with one event per line, ordered
>chronologically. If any of the existing event types match what you
>need, filter for that and combine with the files from other devices. Or
>we can add an event type, I'd probably rather do that than the special
>magic suggested up thread. :)
>
>---
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/syncthing/syncthing/issues/2860#issuecomment-208518440

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by nrm21** [comment]
So I'm still working on this (slowly, off and on) and I have a quick question:

I have a part of the code in model.go where I can determine the delete of file or folder from the local HDD and from that I can just use "m.devicename" to get the friendly name (the one users will want too see, not the long string of hyphenated groups of 7 characters).  But I also have another spot in walk.go where I can determine the create or modify of a file/folder.  However in that code I don't have the 'model' struct available.  I can use "w.ShortID.String()" but that won't give me the friendly name, only the first 7 digits of the long device ID instead.  Whats the easiest way get the friendly name in walk.go?

Also the way I have this planned out is to have it write to a file called **.globallog** in the base dir of whatever share is in question.  So there will be a global log per share.  This way file adds/moves/deletes can take advantage of sync builtin and it will already separate changes between shares instead of dumping all changes into one big massive file.

Only issue I have now is that it won't finish writing to this log until JUST after sync.  Because of this it takes another 60 seconds (or whatever scan time is) before that global log file is then propagated to the rest of the nodes.  It's not hard to see the problems this can cause,  If someone else makes a change to files in that time... boom, we have log conflicts.

Anyone have any thoughts on how I could avoid this issue?

**Comment by calmh** [comment]
> Whats the easiest way get the friendly name in walk.go?

I don't think you should do that. :) I would suggest hooking into `Model.updateLocals()` where we generate the LocalIndexUpdated event. Currently it's just a list of file names, but you could amend that to be a list filenames-and-what-we-did based on what the latest file state is.

> log conflicts

I don't think you should log to a file in the folder. You'll necessarily always run into situations where device A pulls a couple of files at the same time as device B. There is no way you can count on the log just being updated by one device at a time.

I still think the correct way forward is to extent the audit logging appropriately by tweaking the LocalIndexUpdated event, log to a file in the config directory (but per folder if you like), and then merge these files offline when you want the full picture.

**Comment by nrm21** [comment]
Do you mind if I further split the LocalIndexUpdated event into:
LocalIndexUpdatedLocally and LocalIndexUpdatedRemotely?

Or perhaps LocalIndexUpdatedAddition and LocalIndexUpdatedDeletion instead?

**Comment by AudriusButkevicius** [comment]
Thr things you described make little sense.

**Comment by canton7** [comment]
You might want to expand on those terms: what exactly do you intend the two events to describe, what data would be associated with them, etc?

**Comment by nrm21** [comment]
Sorry, I guers I thought it was self documented (maybe I've been thinking about this too long lol):

Instead of an event that says LocalIndexUpdated (to indicated we have updated our local index), we break it up into 2 more specific events:

LocalIndexUpdatedLocally (to indicated we have updated our local index because of a file add/mod/del that WE have made)
LocalIndexUpdatedRemotely (to indicated we have updated our local index because of a file add/mod/del that another node made and sent to us)

**Comment by nrm21** [comment]
Sooooo...

I'm guessing no one has a problem with it then... since no one is complaining?   :)

**Comment by AudriusButkevicius** [comment]
I am failing to understand why these would help.

**Comment by nrm21** [comment]
Ok then I will leave them out.

**Comment by calmh** [comment]
If you're after changes detected locally and nothing else, I guess you could add a ChangesDetected event fired by the scanner somewhere...

**Comment by AudriusButkevicius** [comment]
So I think this is somewhat addressed now with the pull request nrm21 implemented, yet requires some plumbing by the user itself to collect "transactions" from each node to make up the "global" part.

**Comment by nrm21** [comment]
I am currently working on showing these transactions to the user via a row in the GUI.  That was the ultimate point of adding the API functionality.

Shouldn't take alot of time, I just don't always have time to work on or code so I do things in phases.

**Comment by FlorianMickler** [comment]
Thanks for your work, all and nrm21 especially! Great to see syncthing getting improved. I look forward to the day, we have a user visible improvement in the GUI through this effort.

**Comment by nrm21** [comment]
No problem.  I'm also looking forward to seeing it in new versions as it's a feature I've wanted as well.  I expect that will happen any day now in v0.14.17 since it's merged in the main codebase now?

**Comment by AudriusButkevicius** [comment]
There are a few nasty quicks we probably want to fix and a few high-risk features we want to test before releasing, but yeah soon.
